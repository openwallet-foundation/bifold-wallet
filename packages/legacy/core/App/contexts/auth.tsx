// Dont remove the following import line or the pin check will fail when opening askar waller
import '@hyperledger/aries-askar-react-native'

import 'reflect-metadata'

import { AskarWallet } from '@credo-ts/askar'
import { ConsoleLogger, LogLevel, SigningProviderRegistry } from '@credo-ts/core'
import { useAgent } from '@credo-ts/react-hooks'
import { agentDependencies } from '@credo-ts/react-native'
import React, { createContext, useCallback, useContext, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { DeviceEventEmitter } from 'react-native'

import { EventTypes } from '../constants'
import { DispatchAction } from '../contexts/reducers/store'
import { useStore } from '../contexts/store'
import {
  isBiometricsActive,
  loadWalletSalt,
  loadWalletSecret,
  secretForPIN,
  storeWalletSecret,
  wipeWalletKey,
} from '../services/keychain'
import { WalletSecret } from '../types/security'
import { hashPIN } from '../utils/crypto'
import { migrateToAskar } from '../utils/migration'

export interface AuthContext {
  checkPIN: (PIN: string) => Promise<boolean>
  getWalletCredentials: () => Promise<WalletSecret | undefined>
  walletSecret?: WalletSecret
  removeSavedWalletSecret: () => void
  disableBiometrics: () => Promise<void>
  setPIN: (PIN: string) => Promise<void>
  commitPIN: (useBiometry: boolean) => Promise<boolean>
  isBiometricsActive: () => Promise<boolean>
  rekeyWallet: (oldPin: string, newPin: string, useBiometry?: boolean) => Promise<boolean>
}

export const AuthContext = createContext<AuthContext>(null as unknown as AuthContext)

export const AuthProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
  const [walletSecret, setWalletSecret] = useState<WalletSecret>()
  const [store, dispatch] = useStore()
  const { agent } = useAgent()
  const { t } = useTranslation()

  const setPIN = useCallback(async (PIN: string): Promise<void> => {
    const secret = await secretForPIN(PIN)
    await storeWalletSecret(secret)
  }, [])

  const getWalletCredentials = useCallback(async (): Promise<WalletSecret | undefined> => {
    if (walletSecret && walletSecret.key) {
      return walletSecret
    }

    const { secret, err } = await loadWalletSecret(
      t('Biometry.UnlockPromptTitle'),
      t('Biometry.UnlockPromptDescription')
    )

    DeviceEventEmitter.emit(EventTypes.BIOMETRY_ERROR, err !== undefined)

    if (!secret?.key) {
      setWalletSecret(undefined)
      return
    }

    setWalletSecret(secret)

    return secret
  }, [t, walletSecret])

  const commitPIN = useCallback(
    async (useBiometry: boolean): Promise<boolean> => {
      const secret = await getWalletCredentials()
      if (!secret) {
        return false
      }
      // set did authenticate to true if we can get wallet credentials
      dispatch({
        type: DispatchAction.DID_AUTHENTICATE,
      })
      if (useBiometry) {
        await storeWalletSecret(secret, useBiometry)
      } else {
        // erase wallet key if biometrics is disabled
        await wipeWalletKey(useBiometry)
      }
      return true
    },
    [dispatch, getWalletCredentials]
  )

  const checkPIN = useCallback(
    async (PIN: string): Promise<boolean> => {
      try {
        const secret = await loadWalletSalt()

        if (!secret || !secret.salt) {
          return false
        }

        const hash = await hashPIN(PIN, secret.salt)

        if (!store.migration.didMigrateToAskar) {
          await migrateToAskar(secret.id, hash)
          dispatch({
            type: DispatchAction.DID_MIGRATE_TO_ASKAR,
          })
        }

        // NOTE: a custom wallet is used to check if the wallet key is correct. This is different from the wallet used in the rest of the app.
        // We create an AskarWallet instance and open the wallet with the given secret.
        const askarWallet = new AskarWallet(
          new ConsoleLogger(LogLevel.off),
          new agentDependencies.FileSystem(),
          new SigningProviderRegistry([])
        )
        await askarWallet.open({
          id: secret.id,
          key: hash,
        })

        await askarWallet.close()

        const fullSecret = await secretForPIN(PIN, secret.salt)
        setWalletSecret(fullSecret)
        return true
      } catch (e) {
        return false
      }
    },
    [dispatch, store.migration.didMigrateToAskar]
  )

  const removeSavedWalletSecret = useCallback(() => {
    setWalletSecret(undefined)
  }, [])

  const disableBiometrics = useCallback(async () => {
    await wipeWalletKey(true)
  }, [])

  const rekeyWallet = useCallback(
    async (oldPin: string, newPin: string, useBiometry?: boolean): Promise<boolean> => {
      try {
        // argon2.hash can sometimes generate an error
        const secret = await loadWalletSalt()
        if (!secret) {
          return false
        }
        const oldHash = await hashPIN(oldPin, secret.salt)
        const newSecret = await secretForPIN(newPin)
        const newHash = await hashPIN(newPin, newSecret.salt)
        if (!newSecret.key) {
          return false
        }

        await agent?.wallet.close()
        await agent?.wallet.rotateKey({ id: secret.id, key: oldHash, rekey: newHash })
        await storeWalletSecret(newSecret, useBiometry)
        setWalletSecret(newSecret)
      } catch {
        return false
      }
      return true
    },
    [agent]
  )

  return (
    <AuthContext.Provider
      value={{
        checkPIN,
        getWalletCredentials,
        removeSavedWalletSecret,
        disableBiometrics,
        commitPIN,
        setPIN,
        isBiometricsActive,
        rekeyWallet,
        walletSecret,
      }}
    >
      {children}
    </AuthContext.Provider>
  )
}

export const useAuth = () => useContext(AuthContext)
