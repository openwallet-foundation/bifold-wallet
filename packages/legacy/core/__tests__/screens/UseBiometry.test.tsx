import { render, fireEvent, waitFor } from '@testing-library/react-native'
import React from 'react'
import { AuthContext } from '../../App/contexts/auth'
import UseBiometry from '../../App/screens/UseBiometry'
import { testIdWithKey } from '../../App/utils/testable'
import authContext from '../contexts/auth'
import timeTravel from '../helpers/timetravel'
import { BasicAppContext } from '../helpers/app'
import { Linking } from 'react-native'
import { testDefaultState } from '../contexts/store'
import { StoreProvider } from '../../App/contexts/store'
import { RESULTS, check, request } from 'react-native-permissions'

jest.mock('react-native-permissions', () => require('react-native-permissions/mock'))
const mockedCheck = check as jest.MockedFunction<typeof check>
const mockedRequest = request as jest.MockedFunction<typeof request>

jest.spyOn(Linking, 'openSettings').mockImplementation(() => Promise.resolve())

jest.mock('@react-navigation/elements', () => ({
  Header: jest.fn().mockImplementation(() => {
    const Component = () => null;
    Component.displayName = 'Header';
    return Component;
  }),
  HeaderBackButton: jest.fn().mockImplementation(() => {
    const Component = () => null;
    Component.displayName = 'HeaderBackButton';
    return Component;
  }),
  useHeaderHeight: jest.fn().mockReturnValue(150)
}));

jest.mock('react-native-safe-area-context', () => ({
  useSafeAreaInsets: jest.fn().mockReturnValue({
    top: 25,
    bottom: 25,
    left: 0,
    right: 0
  }),
  SafeAreaView: jest.fn().mockImplementation(({children}) => children)
}));

const customStore = {
  ...testDefaultState,
  preferences: {
    ...testDefaultState.preferences,
    useBiometry: false,
  },
}

describe('UseBiometry Screen', () => {
  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    jest.spyOn(global.console, 'error').mockImplementation(() => {})
  })

  beforeEach(() => {
    jest.clearAllMocks()

    authContext.isBiometricsActive = jest.fn().mockResolvedValue(true)
    customStore.preferences.useBiometry = false
    mockedCheck.mockResolvedValue(RESULTS.UNAVAILABLE)
    mockedRequest.mockResolvedValue(RESULTS.BLOCKED)
  })

  test('Renders correctly when biometry available', async () => {
    authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(true)
    const tree = render(
      <BasicAppContext>
        <AuthContext.Provider value={authContext}>
            <UseBiometry />
        </AuthContext.Provider>
      </BasicAppContext>
    )

    await waitFor(() => {
      timeTravel(1000)
    })

    expect(tree).toMatchSnapshot()
  })

  test('Renders correctly when biometry not available', async () => {
    authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(false)
    const tree = render(
      <BasicAppContext>
        <AuthContext.Provider value={authContext}>
            <UseBiometry />
        </AuthContext.Provider>
      </BasicAppContext>
    )

    await waitFor(() => {
      timeTravel(1000)
    })

    expect(tree).toMatchSnapshot()
  })

  test('Toggles use biometrics ok', async () => {
    authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(true)
    const tree = render(
      <BasicAppContext>
        <AuthContext.Provider value={authContext}>
          <UseBiometry />
        </AuthContext.Provider>
      </BasicAppContext>
    )

    const useBiometryToggle = tree.getByTestId(testIdWithKey('ToggleBiometrics'))

    await waitFor(async () => {
      fireEvent(useBiometryToggle, 'valueChange', true)
    })

    const continueButton = tree.getByTestId(testIdWithKey('Continue'))

    await waitFor(async () => {
      fireEvent(continueButton, 'press')
    })

    expect(useBiometryToggle).not.toBeNull()
    expect(continueButton).not.toBeNull()
    expect(authContext.commitPIN).toBeCalledTimes(1)
    expect(tree).toMatchSnapshot()
  })

  describe('ToggleButton Availability', () => {
    test('ToggleButton is enabled when biometry is available', async () => {
      authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(true)

      const { getByTestId } = render(
        <BasicAppContext>
          <AuthContext.Provider value={authContext}>
            <UseBiometry />
          </AuthContext.Provider>
        </BasicAppContext>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))
      expect(toggleButton.props.accessibilityState.disabled).toBe(false)
    })

    test('ToggleButton is enabled even when biometry is not available', async () => {
      authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(false)

      const { getByTestId } = render(
        <BasicAppContext>
          <AuthContext.Provider value={authContext}>
            <UseBiometry />
          </AuthContext.Provider>
        </BasicAppContext>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))
      expect(toggleButton.props.accessibilityState.disabled).toBe(false)
    })
  })

  describe('ToggleSwitch Behavior', () => {
    test('can turn off biometrics regardless of permission status', async () => {
      // Setup with toggle switch is on,
      // and biometric is not available
      customStore.preferences.useBiometry = true
      authContext.isBiometricsActive = jest.fn().mockResolvedValueOnce(false)

      const { getByTestId } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      expect(toggleButton.props.accessibilityState.checked).toBe(false)
    })

    test('turns on biometrics when permission is GRANTED', async () => {
      mockedCheck.mockResolvedValueOnce(RESULTS.GRANTED)

      const { getByTestId } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      expect(toggleButton.props.accessibilityState.checked).toBe(true)
    })

    test('shows settings popup when permission is UNAVAILABLE', async () => {
      mockedCheck.mockResolvedValueOnce(RESULTS.UNAVAILABLE)

      const { getByTestId, getByText } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      expect(getByText('Biometry.SetupBiometricsTitle')).toBeTruthy()
      expect(getByText('Biometry.SetupBiometricsDesc')).toBeTruthy()
      // Toggle should remain off
      expect(toggleButton.props.accessibilityState.checked).toBe(false)
    })

    test('shows settings popup when permission is BLOCKED', async () => {
      mockedCheck.mockResolvedValueOnce(RESULTS.BLOCKED)

      const { getByTestId, getByText } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      expect(getByText('Biometry.AllowBiometricsTitle')).toBeTruthy()
      expect(getByText('Biometry.AllowBiometricsDesc')).toBeTruthy()
      // Toggle should remain off
      expect(toggleButton.props.accessibilityState.checked).toBe(false)
    })

    test('requests permission when status is DENIED and enables when request is GRANTED', async () => {
      mockedCheck.mockResolvedValueOnce(RESULTS.DENIED)
      mockedRequest.mockResolvedValueOnce(RESULTS.GRANTED)

      const { getByTestId } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      expect(request).toHaveBeenCalledTimes(1)
      expect(toggleButton.props.accessibilityState.checked).toBe(true)
    })
  })

  test('requests permission when status is DENIED and switch stays off when request is BLOCKED', async () => {
    mockedCheck.mockResolvedValueOnce(RESULTS.DENIED)
    mockedRequest.mockResolvedValueOnce(RESULTS.BLOCKED)

    const { getByTestId } = render(
      <StoreProvider initialState={customStore}>
        <BasicAppContext>
          <AuthContext.Provider value={authContext}>
            <UseBiometry />
          </AuthContext.Provider>
        </BasicAppContext>
      </StoreProvider>
    )

    const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

    await waitFor(() => {
      fireEvent(toggleButton, 'press')
    })

    expect(request).toHaveBeenCalledTimes(1)
    // Switch stays off
    expect(toggleButton.props.accessibilityState.checked).toBe(false)
  })

  describe('Settings Popup Behavior', () => {
    test('opens app settings when Open Settings is pressed', async () => {
      // Mock permission check to return BLOCKED to trigger settings popup
      mockedCheck.mockResolvedValueOnce(RESULTS.BLOCKED)

      const { getByTestId, getByText } = render(
        <StoreProvider initialState={customStore}>
          <BasicAppContext>
            <AuthContext.Provider value={authContext}>
              <UseBiometry />
            </AuthContext.Provider>
          </BasicAppContext>
        </StoreProvider>
      )

      // Trigger the settings popup
      const toggleButton = getByTestId(testIdWithKey('ToggleBiometrics'))

      await waitFor(() => {
        fireEvent(toggleButton, 'press')
      })

      // Press the Open Settings button
      const openSettingsButton = getByText('Biometry.OpenSettings')
      await waitFor(() => {
        fireEvent(openSettingsButton, 'press')
      })

      expect(Linking.openSettings).toHaveBeenCalledTimes(1)
    })
  })
})
