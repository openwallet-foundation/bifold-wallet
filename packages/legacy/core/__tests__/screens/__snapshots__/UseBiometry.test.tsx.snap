// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseBiometry Screen Renders correctly when biometry available 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "left",
      "right",
      "bottom",
    ]
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#000000",
        "height": "100%",
        "padding": 20,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <
          style={
            Object {
              "marginBottom": 66,
              "minHeight": 200,
              "minWidth": 200,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "rowGap": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.EnabledText1
        </Text>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.EnabledText2
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
             
            Biometry.Warning
          </Text>
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Biometry.UseToUnlock
          </Text>
        </View>
        <View
          style={
            Object {
              "justifyContent": "center",
            }
          }
        >
          <View
            accessibilityLabel="Biometry.Off"
            accessibilityRole="switch"
            accessibilityState={
              Object {
                "busy": undefined,
                "checked": false,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              Object {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="com.ariesbifold:id/ToggleBiometrics"
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgba(211, 211, 211, 1)",
                  "borderRadius": 25,
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 3,
                  "width": 55,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 20,
                    "height": 25,
                    "justifyContent": "center",
                    "transform": Array [
                      Object {
                        "translateX": 2,
                      },
                    ],
                    "width": 25,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#606060",
                        "fontSize": 15,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Óóç
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "margin": 20,
        "marginTop": "auto",
      }
    }
  >
    <View
      accessibilityLabel="Continue"
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        Object {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#42803E",
          "borderRadius": 4,
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="com.ariesbifold:id/Continue"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              },
              false,
              false,
              false,
            ]
          }
        >
          Continue
        </Text>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    presentationStyle="pageSheet"
    style={
      Object {
        "backgroundColor": "#000000",
      }
    }
    transparent={false}
    visible={false}
  />
</RNCSafeAreaView>
`;

exports[`UseBiometry Screen Renders correctly when biometry not available 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "left",
      "right",
      "bottom",
    ]
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#000000",
        "height": "100%",
        "padding": 20,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <
          style={
            Object {
              "marginBottom": 66,
              "minHeight": 200,
              "minWidth": 200,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "rowGap": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.NotEnabledText1
        </Text>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.NotEnabledText2
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Biometry.UseToUnlock
          </Text>
        </View>
        <View
          style={
            Object {
              "justifyContent": "center",
            }
          }
        >
          <View
            accessibilityLabel="Biometry.Off"
            accessibilityRole="switch"
            accessibilityState={
              Object {
                "busy": undefined,
                "checked": false,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              Object {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="com.ariesbifold:id/ToggleBiometrics"
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgba(211, 211, 211, 1)",
                  "borderRadius": 25,
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 3,
                  "width": 55,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 20,
                    "height": 25,
                    "justifyContent": "center",
                    "transform": Array [
                      Object {
                        "translateX": 2,
                      },
                    ],
                    "width": 25,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#606060",
                        "fontSize": 15,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Óóç
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "margin": 20,
        "marginTop": "auto",
      }
    }
  >
    <View
      accessibilityLabel="Continue"
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        Object {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#42803E",
          "borderRadius": 4,
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="com.ariesbifold:id/Continue"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              },
              false,
              false,
              false,
            ]
          }
        >
          Continue
        </Text>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    presentationStyle="pageSheet"
    style={
      Object {
        "backgroundColor": "#000000",
      }
    }
    transparent={false}
    visible={false}
  />
</RNCSafeAreaView>
`;

exports[`UseBiometry Screen Toggles use biometrics ok 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "left",
      "right",
      "bottom",
    ]
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#000000",
        "height": "100%",
        "padding": 20,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <
          style={
            Object {
              "marginBottom": 66,
              "minHeight": 200,
              "minWidth": 200,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "rowGap": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.EnabledText1
        </Text>
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "normal",
            }
          }
        >
          Biometry.EnabledText2
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
             
            Biometry.Warning
          </Text>
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flexShrink": 1,
              "justifyContent": "center",
              "marginRight": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Biometry.UseToUnlock
          </Text>
        </View>
        <View
          style={
            Object {
              "justifyContent": "center",
            }
          }
        >
          <View
            accessibilityLabel="Biometry.Off"
            accessibilityRole="switch"
            accessibilityState={
              Object {
                "busy": undefined,
                "checked": false,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              Object {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="com.ariesbifold:id/ToggleBiometrics"
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgba(211, 211, 211, 1)",
                  "borderRadius": 25,
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "padding": 3,
                  "width": 55,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#FFFFFF",
                    "borderRadius": 20,
                    "height": 25,
                    "justifyContent": "center",
                    "transform": Array [
                      Object {
                        "translateX": 2,
                      },
                    ],
                    "width": 25,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#606060",
                        "fontSize": 15,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Óóç
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "margin": 20,
        "marginTop": "auto",
      }
    }
  >
    <View
      accessibilityLabel="Continue"
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        Object {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(53, 130, 63, 0.35)",
          "borderRadius": 4,
          "opacity": 1,
          "padding": 16,
        }
      }
      testID="com.ariesbifold:id/Continue"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "transform": Array [
                Object {
                  "rotate": "0deg",
                },
              ],
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 25,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Óóï
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              },
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "bold",
                "textAlign": "center",
              },
              false,
              false,
            ]
          }
        >
          Continue
        </Text>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    presentationStyle="pageSheet"
    style={
      Object {
        "backgroundColor": "#000000",
      }
    }
    transparent={false}
    visible={false}
  />
</RNCSafeAreaView>
`;
